### ISSUE TITLE
I appreciate the author's contribution and deeply agree with their approach. However, upon reviewing the article and code, I noticed: Regarding Arm Specific Decoders, is there any discrepancy between the source code and the paper's description?

### ISSUE THREAD

**Issue by yuegu-bingyuan** [Soltanilara/InterACT-LeRobot#3]
code：
class ACTDecoderLayer(nn.Module):
    def __init__(self, config: InterACTConfig):
        super().__init__()
        self.self_attn = nn.MultiheadAttention(config.dim_model, config.n_heads, dropout=config.dropout)
        self.multihead_attn = nn.MultiheadAttention(config.dim_model, config.n_heads, dropout=config.dropout)

        # Feed forward layers.
        self.linear1 = nn.Linear(config.dim_model, config.dim_feedforward)
        self.dropout = nn.Dropout(config.dropout)
        self.linear2 = nn.Linear(config.dim_feedforward, config.dim_model)

        self.norm1 = nn.LayerNorm(config.dim_model)
        self.norm2 = nn.LayerNorm(config.dim_model)
        self.norm3 = nn.LayerNorm(config.dim_model)
        self.dropout1 = nn.Dropout(config.dropout)
        self.dropout2 = nn.Dropout(config.dropout)
        self.dropout3 = nn.Dropout(config.dropout)

        self.activation = get_activation_fn(config.feedforward_activation)
        self.pre_norm = config.pre_norm

    def maybe_add_pos_embed(self, tensor: Tensor, pos_embed: Tensor | None) -> Tensor:
        return tensor if pos_embed is None else tensor + pos_embed

    def forward(
        self,
        x: Tensor,
        encoder_out: Tensor,
        decoder_pos_embed: Tensor | None = None,
        encoder_pos_embed: Tensor | None = None,
    ) -> Tensor:
        """
        Args:
            x: (Decoder Sequence, Batch, Channel) tensor of input tokens.
            encoder_out: (Encoder Sequence, B, C) output features from the last layer of the encoder we are
                cross-attending with.
            decoder_pos_embed: (ES, 1, C) positional embedding for keys (from the encoder).
            encoder_pos_embed: (DS, 1, C) Positional_embedding for the queries (from the decoder).
        Returns:
            (DS, B, C) tensor of decoder output features.
        """
        skip = x
        if self.pre_norm:
            x = self.norm1(x)
        q = k = self.maybe_add_pos_embed(x, decoder_pos_embed)
        x = self.self_attn(q, k, value=x)[0]  # select just the output, not the attention weights
        x = skip + self.dropout1(x)
        if self.pre_norm:
            skip = x
            x = self.norm2(x)
        else:
            x = self.norm1(x)
            skip = x
        x = self.multihead_attn(
            query=self.maybe_add_pos_embed(x, decoder_pos_embed),
            key=self.maybe_add_pos_embed(encoder_out, encoder_pos_embed),
            value=encoder_out,
        )[0]  # select just the output, not the attention weights
        x = skip + self.dropout2(x)
        if self.pre_norm:
            skip = x
            x = self.norm3(x)
        else:
            x = self.norm2(x)
            skip = x
        x = self.linear2(self.dropout(self.activation(self.linear1(x))))
        x = skip + self.dropout3(x)
        if not self.pre_norm:
            x = self.norm3(x)
        return x
paper's description：
Algorithm 2: Multi-arm Decoder
1: Given: Target tokens with positional embeddings Tpos, number of layers L.
2: Let Darm1, Darm2 represent decoders for Arm1 and Arm2 respectively.
3: Let Dsync represent Synchronization block.
4: Initialize cross-attention blocks for Darm1 and Darm2
5: Initialize synchronization block (multihead self-attention)
6: Arm1 Specific Decoder:
7: for each layer l = 1, 2, ..., L do
8: Arm1input ← {Sarm1, CLSarm2, CLSvisual, Svisual}
9: Arm1output ← Darm1(Arm1input, Tpos)
10: end for
11: Arm2 Specific Decoder:
12: for each layer l = 1, 2, ..., L do
13: Arm2input ← {CLSarm1, Sarm2, CLSvisual, Svisual}
14: Arm2output ← Darm2(Arm2input, Tpos)
15: end for
16: Synchronization Block:
17: Concatenatedoutput ← {Arm1output, Arm2output}
18: Sharedoutput ← Dsync(Concatenatedoutput)
19: Split Shared Output:
20: Sharedoutput arm1, Sharedoutput arm2 ← Split(Sharedoutput)
21: Arm Specific Decoder
22: Arm1final output ← Darm1(Arm1input, Sharedoutput arm1)
23: Arm2final output ← Darm2(Arm2input, Sharedoutput arm2)
24: Return Arm1final output, Arm2final output
How are the Arm-specific decoder and synchronization module specifically implemented in the code?

decoder_in = torch.zeros(
            (self.config.chunk_size, batch_size, self.config.dim_model),
            dtype=encoder_in_pos_embed.dtype,
            device=encoder_in_pos_embed.device,
        ) # 

Why initialize with a zero vector (decoder_in)?

**Comment by ivel-Li** [comment]
I also hold there same doubt, and it seems that they discard almost all the encoder out information but leave [cls] tokens only to get decoder output.

**Comment by xndrewlee** [comment]
@yuegu-bingyuan @ivel-Li Sorry for the late response. The original paper was not implemented using LeRobot and it was adapted to it later on. This repository includes many updates made with a revised version of the paper in mind, which ultimately got postponed due to other ongoing projects. I will share a new repo with the paper implementation shortly, sorry for the confusion.

**Comment by xndrewlee** [comment]
@yuegu-bingyuan @ivel-Li I have updated the code (45ad7c69c010721e27b2ada08cf99481f9c7809c), but yet to test thoroughly out on lerobot.

Regarding zero initialization of `decoder_in`:
This is following a standard transformer practice. The zeros are just initialization placeholders and they immediately get positional embeddings added (decoder_in + decoder_pos_embed) which provide the actual query semantics. The real information comes from cross-attention with encoder outputs in the decoder layers.
